demucs/api.py

```diff
@@ -0,0 +1,419 @@
+            more information, please see the Callback section.
+        progress: If true, show a progress bar.
+
+        Callback
+        --------
+        The function will be called with only one positional parameter whose type is `dict`. The
+        `callback_arg` will be combined with information of current separation progress. The
+        progress information will override the values in `callback_arg` if same key has been used.
+
+        Progress information contains several keys (These keys will always exist):
+        - `model_idx_in_bag`: The index of the submodel in `BagOfModels`. Starts from 0.
+        - `shift_idx`: The index of shifts. Starts from 0.
+        - `segment_offset`: The offset of current segment. If the number is 441000, it doesn't
+            mean that it is at the 441000 second of the audio, but the "frame" of the tensor.
+        - `state`: Could be `"start"` or `"end"`.
+        - `audio_length`: Length of the audio (in "frame" of the tensor).
+        - `models`: Count of submodels in the model.
+        """
+        self._set_attr("_device", device)
+        self._set_attr("_shifts", shifts)
+        self._set_attr("_overlap", overlap)
+        self._set_attr("_split", split)
+        self._set_attr("_segment", segment)
+        self._set_attr("_jobs", jobs)
+        self._set_attr("_progress", progress)
+        self._set_attr("_callback", callback if callable(callback) else None)
+        self._set_attr("_callback_arg", callback_arg)
+
+    def _load_model(self):
+        self._model = get_model(name=self._name, repo=self._repo)
+        if self._model is None:
+            raise LoadModelError("Failed to load model")
+        self._audio_channels = self._model.audio_channels
+        self._samplerate = self._model.samplerate
+
+    def _load_audio(self, track: Path):
+        errors = {}
+        wav = None
+
+        try:
+            wav = AudioFile(track).read(streams=0, samplerate=self._samplerate,
+                                        channels=self._audio_channels)
+        except FileNotFoundError:
+            errors["ffmpeg"] = "FFmpeg is not installed."
+        except subprocess.CalledProcessError:
+            errors["ffmpeg"] = "FFmpeg could not read the file."
+
+        if wav is None:
+            try:
+                wav, sr = ta.load(str(track))
+            except RuntimeError as err:
+                errors["torchaudio"] = err.args[0]
+            else:
+                wav = convert_audio(wav, sr, self._samplerate, self._audio_channels)
+
+        if wav is None:
+            raise LoadAudioError(
+                "\n".join(
+                    "When trying to load using {}, got the following error: {}".format(
+                        backend, error
+                    )
+                    for backend, error in errors.items()
+                )
+            )
+        return wav
+
+    def separate_tensor(self, wav: th.Tensor) -> Tuple[th.Tensor, Dict[str, th.Tensor]]:
```

Comment:
should we force people to pass in the sample rate as well ? I would even put the sample rate as mandatory, because most people don't know what it is and will forget to pass it if they are not forced to !

Reply:
Maybe it's a good idea